whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,414 @@
+You may come across a programming task where you want to implement something like:
+
+Try _this_ and, if it doesn't work out, try _that_ - or one could say a computation composed of two (or more) alternatives which should be tried
+one after the other until one of them produces a useful result.
+
+As an example, lets build a function which takes a `String` as input and returns `Just String` if that input is either `"bob"` or `"alice"`.
+
+Although there might be a lot of other ways to solve this task, we'll do it by defining two functions:
+
+```haskell
+isAlice :: String -> Maybe String
+isAlice a = if a == "alice" then Just a else Nothing
+
+isBob :: String -> Maybe String
+isBob a = if a == "bob" then Just a else Nothing
+```
+
+We can then compose these two functions like this:
+
+```haskell
+isAliceOrBob :: String -> Maybe String
+isAliceOrBob a =
+  case isAlice a of
+    Just _ -> Just a
+    Nothing -> case isBob a of
+                 Just _ -> Just a
+                 Nothing -> Nothing
+```
+
+```none
+λ> isAliceOrBob "alice"
+Just "alice"
+λ> isAliceOrBob "bob"
+Just "bob"
+λ> isAliceOrBob "charly"
+Nothing
+```
+
+This works well, but it will become ugly when we add more alternatives to the computation  f.e. a `isCharly :: String -> Maybe String`.
+
+It would be nice if we could have an operator to write it in a nice way like this:
+
+```haskell
+isCharly :: String -> Maybe String
+isCharly a = if a == "charly" then Just a else Nothing
+
+isAliceOrBobOrCharly :: String -> Maybe String
+isAliceOrBobOrCharly a = isAlice a <|> isBob a <|> isCharly a
+-- or
+isAliceOrBobOrCharly a = isAliceOrBob a <|> isCharly a
+```
+
+
+## The Alternative typeclass
+
+Actually, the code above is valid Haskell. All you need to do to make use of the `<|>` operator is to import it from `Control.Aplicative`:
+
+```none
+λ> :info (<|>)
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  ...
+  (<|>) :: f a -> f a -> f a
+  ...
+  	-- Defined in ‘GHC.Base’
+infixl 3 <|>
+```
+
+It's a method of the `Alternative` typeclass. Let's look at what information GHCI can provide us about this typeclass:
+
+```haskell
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  empty :: f a
+  (<|>) :: f a -> f a -> f a
+  some :: f a -> f [a]
+  many :: f a -> f [a]
+  {-# MINIMAL empty, (<|>) #-}
+  	-- Defined in ‘GHC.Base’
+instance Alternative ZipList -- Defined in ‘Control.Applicative’
+instance Alternative [] -- Defined in ‘GHC.Base’
+instance Alternative Maybe -- Defined in ‘GHC.Base’
+instance Alternative IO -- Defined in ‘GHC.Base’
+```
+
+This already gives us a lot of useful facts:
+
+1. A type implementing `Alternative` must be of kind `* -> *`.
```

Comment:
Is this article aimed at beginners? It's very likely they don't know what `* -> *` is. We have an article about kinds, might be nice to add a link to it.

Reply:
I think you can expect that the people reading this article are a bit more experienced and know what a kind is. 

(It's always nice to cross-link things, though. https://serokell.io/blog/kinds-and-hkts-in-haskell <-- here's the link)

Reply:
added a link the article.